Add MongoDB GridFS Adapter

GromNaN opened the pull request:
[GridFS](https://www.mongodb.com/docs/manual/core/gridfs/) is a file system that uses MongoDB as storage. I would like to propose this adapter which uses the [MongoDB PHP Library](https://github.com/mongodb/mongo-php-library/).

Docs PR: https://github.com/thephpleague/flysystem/pull/1795

The [legacy GridFS Flysystem adapter](https://github.com/thephpleague/flysystem-gridfs) had some [significant number of installations](https://packagist.org/packages/league/flysystem-gridfs/stats), even if it used the legacy Mongo extension. It shows there's some interest.

Important notes:
- Directories does not exist in GridFS. I use the same technic as for S3 when explicitly creating a directory: an empty file with a trailing `/` is created. The creation of a file with a name ending in a `/` is therefore prohibited.
- Permissions are not supported by GridFS. When explicitly set, I store the visibility in the `metadata.flysystem_visibility` attribute of the file.
- GridFS' `contentType` attribute [is deprecated](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.md#design-rationale). The attribute `metadata.contentType` is used instead.
- GridFS supports versioning by default. Writing to an existing file creates a new revision. Deleting removes all the revisions. Moving renames all the revisions.
- For GitHub Action, the [mongodb extension](https://pecl.php.net/package/mongodb) is installed and a mongodb server is started using  Docker Compose.
- I added some tests in the shared `FilesystemAdapterTestCase` in order to improve coverage on `mimeType` and `visibility`.

GromNaN committed:
GridFS Adapter

jmikola commented on src/GridFS/.github/FUNDING.yml:
```diff
@@ -0,0 +1,2 @@
+github: [frankdejonge]
+custom: "https://ecologi.com/frankdejonge"
```
Is this file relevant? It doesn't appear in other providers under `src/`: https://github.com/thephpleague/flysystem/tree/3.x/src

GromNaN replied:
Removed. I don't know how it landed here.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,404 @@
+        try {
+            $stream = $this->bucket->openDownloadStream($file['_id']);
+            $this->bucket->uploadFromStream($this->prefixer->prefixPath($destination), $stream, $options);
+        } catch (Exception $exception) {
```
Just confirming: this is intended to catch _any_ exception from PHPC _or_ PHPLIB (noting the PHPLIB exceptions all extend the PHPC interface as well). Basically, anything from this provider gets proxied by a Flysystem exception.

GromNaN replied:
Yes, flysystem adapters wrap all the exceptions.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,404 @@
+        }
+
+        try {
+            $this->bucket->getFilesCollection()->updateMany(
```
Noted that this is the same approach used in https://github.com/mongodb/mongo-php-library/pull/1207 ([PHPLIB-1324](https://jira.mongodb.org/browse/PHPLIB-1324)), despite there being [no standard approach documented in the GridFS spec](https://jira.mongodb.org/browse/DRIVERS-2807).

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,404 @@
+        $pipeline = [];
+        if ($path !== '') {
+            $pathdeep = substr_count($path, '/');
+            // Exclude files that does not start with the expected path
```
```suggestion
            // Exclude files that do not start with the expected path
```

Subject/verb agreement.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,404 @@
+                }
+            }
+        } else {
+            // Get the last revision of each file name
```
```suggestion
            // Get the metadata of the last revision of each file
```

For consistency with the earlier pipeline construction. I assume these should be consistent.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    public function createDirectory(string $path, Config $config): void
```
Is this the same approach used in other storage drivers that don't have a native concept of directories (e.g. S3)? I wonder if it'd be sensible to add a boolean metadata field to identify this as a directory, as there is otherwise nothing to distinguish this from an empty file.

I'm thinking more about other tooling that may work with the same GridFS bucket, even if such a field isn't directly usable by this adapter.

GromNaN replied:
Using this adapter, trailing `/` are forbidden. So a file with a trailing `/` is a directory.

I just added a `metadata.flysystem_directory` as marker.

jmikola commented on src/GridFS/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2013-2023 Frank de Jonge
```
```suggestion
Copyright (c) 2024-present Frank de Jonge
```

Assuming this was copied from elsewhere.

jmikola commented on src/GridFS/composer.json:
```diff
@@ -0,0 +1,21 @@
+        "php": "^8.0.2",
+        "ext-mongodb": "^1.15",
+        "league/flysystem": "^3.10.0",
+        "mongodb/mongodb": "^1.15"
```
Is there something particular in the [1.15.0 release](https://jira.mongodb.org/secure/ReleaseNote.jspa?projectId=12483&version=34330) that is required by this adapter, or was this an arbitrary choice?

GromNaN replied:
Nothing specific. I didn't have the courage to go looking for the oldest version that supported the GridFS features required.

GromNaN replied:
I tested version by version, starting with 1.0, and in the end it was quick because 1.2 seems to be compatible, with https://github.com/mongodb/mongo-php-library/commit/a448b609bdf6e9f0a5fb8a9d0cdd8583b6502245

jmikola replied:
The pedant in me appreciates using the lowest necessary dependency, but we'd probably also be fine to stick with whatever overlaps with Flysystem's PHP language requirement. Up to you.

frankdejonge commented:
Hi all, I'm out this week but will get on this ASAP to get this out. Thanks for the effort and the interest to contribute. Would there be anybody interested from your end to do continued maintenance in this adapter? And if so, who? Then I know who to rope in with issues.

GromNaN commented:
Hello Frank, we are a team of currently 3 devs, led by @alcaeus. We use an open [Jira](https://jira.mongodb.org/browse/PHPORM-66?jql=project%20%3D%20PHPORM%20AND%20component%20%3D%20Flysystem) to organise our work and Flysystem is one of the packages we want to maintain.

frankdejonge commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,413 @@
+
+    public function read(string $path): string
+    {
+        return stream_get_contents($this->readStream($path));
```
in some PHP environments, the stream opened is not automatically closed, an implementation like to following is needed:

```php
    public function read(string $path): string
    {
        $readStream = $this->readStream($path);
        $contents = stream_get_contents($readStream);
        fclose($readStream);

        return $contents;
    }
```

GromNaN replied:
Updated to use `try...finally`

frankdejonge commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,413 @@
+        }
+
+        // Deleting a file that does not exist is no-op
+        while ($file = $this->findFile($path)) {
```
1) why is this a while loop?
2) this block may be affected by a race condition when two deletions are happening concurrently. In other implementations we try to delete and ignore the exception if it was caused by a "file not found" situation. Can we inspect the exception to see if occurs and ignore the error when that's the case?

GromNaN replied:
The while was necessary to remove all the revisions of the file name. I replaced the code with a "find all revisions" followed by "delete" on each revision. Good point about race-condition, I added a try-catch.

frankdejonge commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,413 @@
+            throw UnableToRetrieveMetadata::mimeType($path, 'file does not exist');
+        }
+
+        return new FileAttributes($path, null, $file['metadata'][self::METADATA_VISIBILITY] ?? null);
```
the last constructor parameter is an "extra metadata" field, I'd recommend putting the entire metadata value there

frankdejonge replied:
you can also map everything in one go, defaulting values to null in a central `metadata` method and only check if the value is there in the corresponding getter, this way you map once and check for expected values per case. WDYT?

frankdejonge commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,413 @@
+                    );
+                } else {
+                    $file = $file['file'];
+                    yield new FileAttributes(
```
this mapping, if extracted into a method could be the implementation of the before mentioned `metadata` method

GromNaN replied:
Good idea, I factorized it with the `mapFileAttributes` function.

GromNaN committed:
Always map all attrributes

Fix ext-mongodb explicit dependency for exception

Use generic type for file object

GromNaN committed:
Factorize delete operation and ignore missing file ignore due to race-conditions

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+        if ($visibility = $config->get(Config::OPTION_VISIBILITY)) {
+            $options['metadata'][self::METADATA_VISIBILITY] = $visibility;
+        }
+        if (($mimetype = $config->get('mimetype')) || ($mimetype = $this->mimeTypeDetector->detectMimeTypeFromPath($path))) {
```
I noticed this was done in `write()` as well, but what's the point of inferring a mime type from the destination filename? Shouldn't this be inferred from the actual contents, like [file(1)](https://man7.org/linux/man-pages/man1/file.1.html)?

Maybe this doesn't matter at all since `detectMimeTypeFromPath()` is only fall-back behavior if an explicit `mimetype` option isn't specified. I'm assuming the Config object is a container for per-operation options and not some common configuration for the entire library.

GromNaN replied:
The content can only be used when the file is created with `write`. The methods `writeStream` and `copy` can't access the contents of the file, only its name.
We could also use fallback only when reading, but in the same way we don't know the contents of the file when the metadata are read.

Yes, `Config` is set per operation and contains the metadata `mimetype` and `visibility`.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+        }
+
+        try {
+            return $this->bucket->openDownloadStream($file['_id']);
```
Why not [`openDownloadStreamByName()`](https://www.mongodb.com/docs/php-library/current/reference/method/MongoDBGridFSBucket-openDownloadStreamByName/)?

GromNaN replied:
Good point, I've changed it.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+        ];
+
+        if ($visibility = $config->get(Config::OPTION_VISIBILITY)) {
+            $options['metadata'][self::METADATA_VISIBILITY] = $visibility;
```
What's the purpose of this option? I don't follow why it's only injected into `metadata` when truthy. I noted this same logic appears earlier in the `write` and `writeStream` methods.

It seems contrary to the logic in `setVisibility()` below, which could end up storing a true _or_ false value to the field.

GromNaN replied:
I don't want to store visibility unless it's explicitly set. `setVisibility` requires a value. For `write`, `writeStream`, `createDirectory` and `copy`, this is an option.

@frankdejonge Since there's no concept of file-level permission in GridFS, would it be better to disable the functionality altogether, as is done for WebDAV?

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+            throw UnableToRetrieveMetadata::mimeType($path, 'unknown');
+        }
+
+        return new FileAttributes($path, null, null, null, $file['metadata'][self::METADATA_MIMETYPE]);
```
Should this use `mapFileAttributes()`?

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+            $pipeline[] = ['$match' => ['filename' => new Regex('^' . preg_quote($path))]];
+        }
+        // Get the last revision of each file
+        $pipeline[] = ['$sort' => ['filename' => 1, 'uploadDate' => -1]];
```
Is `listContents()` always expected to sort on the filename? Just curious since this isn't mentioned in the comment.

GromNaN replied:
I don't think there's any specification on this point. The idea here is to make full use of [the index](https://github.com/mongodb/mongo-php-library/blob/cbc8104c0b2c32b7cf572ff759324c872e8dc63a/src/GridFS/CollectionWrapper.php#L324) on `filename` and `uploadDate`.

jmikola replied:
Noted. Thanks for adding a comment.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+                'file' => ['$first' => '$$ROOT'],
+                // The "lastModified" date is the date of the last uploaded file in the directory
+                'uploadDate' => ['$max' => '$uploadDate'],
+            ]];
```
Does this pipeline correctly filter out documents _above_ the target directory?

Consider:

```
> db.foo.insertOne({filename:'/a/file.txt'})
> db.foo.insertOne({filename:'/a/b/file.txt'})
> db.foo.insertOne({filename:'/a/b/c/file.txt'})
> db.foo.aggregate([
  {$addFields:{splitpath:{$split:['$filename','/']}}},
  {$group:{
    _id:{
      basename:{$arrayElemAt:['$splitpath',3]},
      isDir:{$ne:[{$size:'$splitpath'},4]}
    },
    file: {'$first': '$$ROOT'}
  }}
])
```

The above pipeline uses `$pathdeep` of 3 and yields the following:

```
[
  {
    _id: { isDir: true },
    file: {
      _id: ObjectId("664b625ca81bc8b8d615421c"),
      filename: '/a/file.txt',
      splitpath: [ '', 'a', 'file.txt' ]
    }
  },
  {
    _id: { basename: 'file.txt', isDir: false },
    file: {
      _id: ObjectId("664b6260a81bc8b8d615421d"),
      filename: '/a/b/file.txt',
      splitpath: [ '', 'a', 'b', 'file.txt' ]
    }
  },
  {
    _id: { basename: 'c', isDir: true },
    file: {
      _id: ObjectId("664b6263a81bc8b8d615421e"),
      filename: '/a/b/c/file.txt',
      splitpath: [ '', 'a', 'b', 'c', 'file.txt' ]
    }
  }
]
```

I was consistent about using a leading `/` in the filenames here, but I suppose that could vary in practice (especially if Flysystem isn't used exclusively for IO on the bucket).

GromNaN replied:
There is a `$match` stage to filter-out files that are not in the specified directory.
```ts
db.foo.aggregate([
  {$match:{filename:/^\/a\/b\/c\//}},
  {$addFields:{splitpath:{$split:['$filename','/']}}},
  {$group:{
    _id:{
      basename:{$arrayElemAt:['$splitpath',3]},
      isDir:{$ne:[{$size:'$splitpath'},4]}
    },
    file: {'$first': '$$ROOT'}
  }}
])
```

```ts
[
  {
    _id: { basename: 'c', isDir: true },
    file: {
      _id: ObjectId("664b706bf6dc6882f4ae8b93"),
      filename: '/a/b/c/file.txt',
      splitpath: [ '', 'a', 'b', 'c', 'file.txt' ]
    }
  }
]
```

jmikola replied:
Thanks. Missed that when testing locally.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+        $pathdeep = 0;
+        $pipeline = [];
+        if ($path !== '') {
+            $pathdeep = substr_count($path, '/');
```
Does this function behave consistently irrespective of a trailing `/` in `$path`?

GromNaN replied:
`PathPrefixer` adds the trailing slash: https://github.com/thephpleague/flysystem/blob/0487c25a61e9b17ce82559af9bc1fc7934ee53b0/src/PathPrefixer.php#L48
So we always have a leading and a trailing slash.

jmikola commented on src/GridFS/GridFSAdapter.php:
```diff
@@ -0,0 +1,443 @@
+class GridFSAdapter implements FilesystemAdapter
+{
+    private const METADATA_DIRECTORY = 'flysystem_directory';
+    private const METADATA_VISIBILITY = 'flysystem_visibility';
```
What's the purpose of this metadata field? I only see it being set during write operations and incorporated into returned FileAttributes or DirectoryAttributes.

GromNaN replied:
Related to https://github.com/thephpleague/flysystem/pull/1794#discussion_r1606864392

GromNaN committed:
Use openDownloadStreamByName

GromNaN committed:
Use mapFileAttributes in mimeType

GromNaN committed:
Use file index in the specified order

jmikola reviewed (approved):
LGTM. I'll defer to you both to decide whether to remove the the visibility option (https://github.com/thephpleague/flysystem/pull/1794/files#r1606938015).

frankdejonge commented:
Looks like the test case `GridFSAdapterTest::move_all_revisions` is failing.

GromNaN commented:
> Looks like the test case `GridFSAdapterTest::move_all_revisions` is failing.

Where do you see that? It seems to work in GitHub Actions: https://github.com/thephpleague/flysystem/actions/runs/9161336038/job/25242807530?pr=1794

GromNaN committed:
Update package name

frankdejonge commented:
@GromNaN this is an example; https://github.com/thephpleague/flysystem/actions/runs/9180056390/job/25244027378?pr=1794 might be because of eventual consistency in MongoDB? Is that a thing? Previous run had more errors on the same test case.

GromNaN committed:
Ensure files are created in a different millisecond in test to keep revisions order

GromNaN commented:
Ok, I suspect that it's because 2 files are created in the same millisecond and revisions are sorted using this date.

GromNaN committed:
Set readPreference in tests

GromNaN commented:
I updated the `readPreference`, so it should be OK.

frankdejonge commented:
@GromNaN still one failing, is there another way to deal with this or to configure it? https://github.com/GromNaN/flysystem/actions/runs/9181466558/job/25248713035

GromNaN committed:
Remove read preference, as primary is the default

GromNaN commented:
It seems that my previous guess was correct but the fix incorrectly implemented. I added `usleep(1000)` (1ms) between each insertion of a revision to have a different `uploadDate`.
